### ISSUE TITLE
Stacked bars collapsed

### ISSUE THREAD

**Issue by ClaraMedopad** [ecomfe/echarts#7937]
### One-line summary [问题简述]

Stacked bar charts collapsed since 4.0.3. See screenshot. 

### Version & Environment [版本及环境]
+ ECharts version [ECharts 版本]: 4.0.4
+ Browser version [浏览器类型和版本]: Chrome 64
+ OS Version [操作系统类型和版本]:  macOS High Sierra

### Expected behaviour [期望结果]

Bars able to stack on top of each other without overflowing issues.

### ECharts option [ECharts配置项]

```javascript
option = { 
  grid: {
    right: '8%',
    left: '8%'
  },
  xAxis: {
      type: 'time',
      minInterval: 1000 * 3600 * 24
  },
  yAxis: {
    type: 'value',
    min: 0,
    max: 100,
    name: '%'
  },
  series: [
  {
    type: 'bar',
    stack: 'adherence',
   data: [{
       value: ['2018-02-21', 6, 2, 33, 67]
     }],
    encode: { x: 0, y: 3 },
    barMaxWidth: 25,
    barCategoryGap: '50%',
    cursor: 'auto',
    itemStyle: {
      normal: {
        color: 'red',
        opacity: 0.5
      }
    }
  },
  {
    type: 'bar',
    stack: 'adherence',
    data: [{
       value: ['2018-02-21', 6, 2, 33, 67]
     }],
    encode: { x: 0, y: 4 },
    barMaxWidth: 25,
    barCategoryGap: '50%',
    cursor: 'auto',
    itemStyle: {
      normal: {
        color: 'green',
        opacity: 0.5
      }
    }
  }],
  dataZoom: {
    type: 'slider',
    showDataShadow: false
  }
}
```

### Other comments [其他信息]

![screen shot 2018-03-14 at 14 44 01](https://user-images.githubusercontent.com/36439799/37409504-657fb018-2796-11e8-8b32-34717e6d275d.png)

**Comment by ClaraMedopad** [comment]
Switching x-axis type to 'category' solved the problem. But unsure why issue only arose after version upgrade from 4.0.2 to 4.0.4.

**Comment by ClaraMedopad** [comment]
Reopened because I need to use time for x-axis 😞

**Comment by ClaraMedopad** [comment]
Some more information..
In src/data/helper/dataStackHelper.js: 

``` 
export function isDimensionStacked(data, stackedDim, stackedByDim) {
return stackedDim
        && stackedDim === data.getCalculationInfo('stackedDimension')
        && (
            stackedByDim != null
                ? stackedByDim === data.getCalculationInfo('stackedByDimension')
                : data.getCalculationInfo('isStackedByIndex')
        );
}
```
stackedByDim is x and data.getCalculationInfo('stackedByDimension') is undefined.

**Comment by PicusViridis** [comment]
I have the same problem, stacked bar chart with time x-axis worked in 4.0.2 but not until 4.0.3.

**Comment by 100pah** [comment]
Currently, `stack` is only supported on `'category'` axis.

In fact, we've been considering whether support `stack` on `'time'` or event `'value'` axis, since it is a bit complicated than supporting it on `'category'` axis.

Could I have the detailed reason that why the `'time'` axis is necessary in your case?

**Comment by ClaraMedopad** [comment]
In my case I use it to visualise medication adherence, i.e. how many tablets a patient has taken out of the total per day. I want to have a continuous overview of the data, i.e. if the medication is stopped or not required for the day, a gap will be shown in the bar chart.

**Comment by BenoitHanotte** [comment]
Same issue here, I use echarts to visualize memory consumption over time and I stack memory for different applications or part of the memory along with time (eg: heap memory + off-heap memory = total memory consumed, thus I want to stack series)

**Comment by Xalio08** [comment]
Same issue, I need to stack registrations of different types at different dates, and categories doesn't work since some days are empty and should still appear.

**Comment by tmtron** [comment]
our use-case is almost like the [Confidence Band example](https://ecomfe.github.io/echarts-examples/public/editor.html?c=confidence-band), but:

- but for the line-series, we need the time axis
- for the confidence band we could use a category axis


any ideas for a workaround? e.g. 

1. maybe we can have 2 x-axis: a time-axis for the line-series and a category axis for the (confidence band stack)? 
But those axis would need to be "sychronized": e.g. when we zoom in, both must zoom together, etc.
1. or is it worth to try to keep only one time-axis and use a [custom series](https://ecomfe.github.io/echarts-doc/public/en/option.html#series-custom.type) to draw the area?
If so, can someone link to an example, how to use custom-series with a time x-axis?
1. something completely different, that I didn't think of..

**Comment by tmtron** [comment]
> In fact, we've been considering whether support `stack` on `'time'` or event `'value'` axis, since it is a bit complicated than supporting it on `'category'` axis.

@100pah: will it be simpler to support only a special case and not the general case to stack 2 different series (which may have data-samples at different time-stamps)?

e.g. when you have measurement data, you often need to display the min and max values of a series (instead of just e.g. the average) to avoid aliasing: e.g. in the screenshot below we have one such series (1) is the maximum and (2) is the minimum: so each of the samples have a time-stamp and they always have min and max:
![image](https://user-images.githubusercontent.com/20422904/60463411-36a8f200-9c4c-11e9-8371-9e0c86dbdc8c.png)

**Comment by tmtron** [comment]
+1 don't close

**Comment by ciesielskico** [comment]
+1, we need stacking with x-axis type `value`

**Comment by hughess** [comment]
+1, I assume same solution to #15102

**Comment by th1under** [comment]
+1, It's related to #16177 and #16099

We also need stacking bars on time xAxis like the example provided in #16177.

**Comment by ermaugustinas** [comment]
+1 Still broken, we're interested to use stacked bar with time.

**Comment by rivaldi8** [comment]
We have an issue with stacked column too. They appear floating:

![stacked-columns-bad](https://user-images.githubusercontent.com/1187977/225908723-ca2f1b45-89fb-40b2-8cd1-04de906442b7.png)

I'll open a separate issue later when I find some time. I've run a bisect and found the issue was introduced in 20eaa3d56e62f1afc14675c75120683389e7ebfe, in case you want to check if it affects this issue too.

**Comment by juliepagano** [comment]
Has something changed that makes this slightly more supported now or are there still limitations on stacking on a time axis? When I [try the options provided by the original poster](https://echarts.apache.org/examples/en/editor.html?code=PYBwLglsB2AEC8tSRg2BvWAoWsDmAThACYBcGOusReAFmOQOQAcApIwDSW4A2ApgDMGsFu0oBfLrgAeAQWkQAzuXTcqYAJ4g-TSAFs-nNbj0RoASWhg-BAG4BDHuQCMABnewAVLADMANg9vACYAFgkpWA15JRU1TW0mBx4AV0MIkzNyV3TYPXtpF3cc6HsDJjFcSUpFGwg-ZVgAbUpVKlh4nREAI3sCIzbFMHsAYwBrJntiWhs-aGG07mJ7IfJG1rbcJNTVxiDXZ2YAWlcgw6DnTlg_Dlggm58fG78AdgBdY3FXnNnh4GJOzAFWDZSLkHywKptHoEACy-QA6iQwLRyEEAKw5aEAYWWfDwwAIGgA4vYQEw0a52DlhskCIoCRNkmBgP0qBBrHoAMqafixDawaAEvJOCj83C_HgMkQEPjEVn80AjdkaLIAOjRxghanE4RacS0nUY0PlsEGI3GIkm0xlcwWbSWKya6w2W06jV2-yOJzOFyeNzuvkeVzeHy-ah-fwBsCBIJVsBCEMxvTh0kRxGRqIxamxuPxhJJZJEFKpahpdKljHsTJZOXZfC5PIBmsFBGFfLFEorhD4sxNbUVw2Vao1_J1bTHELDuAd9gAWsBgHp27gOkxFDwSDYTYpaMAAO4AEWW9k5tEm-_IAkcNQkWHEQA&_source=echarts-doc-preview&version=5.4.3), it looks like it might be properly stacked now.

<img width="1029" alt="image" src="https://github.com/apache/echarts/assets/396962/41875e9d-eff3-4b96-9f5a-be19541f5daa">

**Comment by mitar** [comment]
It seems this still does not work for time, value, and log. :-(

**Comment by helgasoft** [comment]
@mitar, stacked bars on **time** axis work - [Demo code](https://echarts.apache.org/examples/en/editor.html?c=bar-stack-normalization&code=PYBwLglsB2AEC8sDeBYAULTsAeBBbEAzgFzJgCeIApqQOSQC2VtAvgDTpazn5GlLtOWADZUA5lWgATfrEEYshKgCcIVErADaqBV0zQAhk1IAiAyY669Ug2AOlNmgIxsnAJgAMAXTZa3bT29fTQBmVwBWIK0AFjYADijNcLYAdkSANlcnRJTXEO8fIT1YCmo6ACMDZVpLYsxCOwBjAGs6GGYiuV8dOthDY1gTcotOrhs7B2dXTx8_AI9EsKdI2c1YhNXkp23VzO2cvILautKaWFpK6uPihoMWtugOqxYvdBYAbiA)

**Comment by mitar** [comment]
@helgasoft So this issue can be then closed?

**Comment by gggdomi** [comment]
> @mitar, stacked bars on **time** axis work - [Demo code](https://echarts.apache.org/examples/en/editor.html?c=bar-stack-normalization&code=PYBwLglsB2AEC8sDeBYAULTsAeBBbEAzgFzJgCeIApqQOSQC2VtAvgDTpazn5GlLtOWADZUA5lWgATfrEEYshKgCcIVErADaqBV0zQAhk1IAiAyY669Ug2AOlNmgIxsnAJgAMAXTZa3bT29fTQBmVwBWIK0AFjYADijNcLYAdkSANlcnRJTXEO8fIT1YCmo6ACMDZVpLYsxCOwBjAGs6GGYiuV8dOthDY1gTcotOrhs7B2dXTx8_AI9EsKdI2c1YhNXkp23VzO2cvILautKaWFpK6uPihoMWtugOqxYvdBYAbiA)

It works with your specific data which has all values for all series. I changed a single character (the last datapoint) and [here's the incorrect result](https://echarts.apache.org/examples/en/editor.html?c=bar-stack-normalization&code=PYBwLglsB2AEC8sDeBYAULTsAeBBbEAzgFzJgCeIApqQOSQC2VtAvgDTpazn5GlLtOWADZUA5lWgATfrEEYshKgCcIVErADaqBV0zQAhk1IAiAyY669Ug2AOlNmgIxsnAJgAMAXTZa3bT29fTQBmVwBWIK0AFjYADijNcLYAdkSANlcnRLjXEO8fIT1YCmo6ACMDZVpLYsxCOwBjAGs6GGYiuV8dOthDY1gTcotOrhs7B2dXTx8_AI9EsKdI2c1YhNXkp23VzO3ElLyC2rrSmlhaSuqT4oaDFrboDqsWL3QWAG4gA).

**Comment by helgasoft** [comment]
> It works with your specific data which has all values for all series.

yes @gggdomi. And "incorrect results" are due to missing values which could/should be fixed by data preprocessing. 
[Demo](https://echarts.apache.org/examples/en/editor.html?code=PYBwLglsB2AEC8sDeBYAULTsAeBBbEAzgFzJgCeIApqQOSQC2VtAvgDTpazn5GlLtOWADZUA5lWgATfrEEYshKgCcIVErADaqBV0zQAhk1IAiAyY669Ug2AOlNmgIxsnAJgAMAXTZa3bT29fTQBmVwBWIK0AFjYADijNcLYAdkSANlcnRJS2aABXYWEfLTjXEO8fIT1YCmo6ACMDZVpLGsxCOwBjAGs6GGZquV8ddthDY1gTBoshrhs7B2dXTxLNf0C1sKdItdiEteSnY7XM45zyxLKCoq8qqy46mlhaJpa29s6DXv7oQasWF50CwANxAA) with missing values added.

### FAQ: ECharts axis types
ECharts has four [axis types](https://echarts.apache.org/en/option.html#xAxis.type) available - _value_, _category_, _time_ and _log_.
- category is a **discrete** axis with individual values
- time and value are **continuous** axes with a range of values without gaps
- log is a non-linear axis dealing with very-large-range values

**Character data** are naturally discrete and can only be represented on a _category_ axis. 
Examples:  ```['Peru','Cuba','Japon',...]``` or ```['Mon','Tue','Wed',...]``` or ```['TypeA','TypeB',...]```.
**Numerical data** is usually continuous, like temperature, price or distance. 
**Time data** is also continuous, with resolution in minutes, days or years.
 
All axis types allow for **missing data**, but present them differently. For instance
A bar chart with ```data:[['Feb',22], ['Oct',44]]``` on a **category** axis will present them side-by-side, 
but ```data:[['2024-02',22], ['2024-10',44]]``` on a **time** axis will show continuous time without data in-between.
Missing time data, as null or interpolated values, could be added in a **data pre-processing** step.
Data transformation is a developer's task. Often data has to be modified to meet the needs of a chosen charting library and not the other way around.

**Log axis** is not suitable for stacked charts. A stacked chart relies on a linear axis to show proportions, but log axis is not linear.

---

**Comment by gggdomi** [comment]
@helgasoft thanks for your answer, even if I politely disagree. I guess this issue can be closed, in favor of https://github.com/apache/echarts/issues/20254 which is cleaner.

**Comment by kianj2** [comment]
+1

@helgasoft I appreciate your explanation, as it is the most thorough one I have found in this thread. I personally do not feel it addresses the comments made by @gggdomi. I see your point that preprocessing needs to be done, but that is mentioned no where in the documentation nor do I feel that is implied when using a time axis. In your example playground, we need to add `[7,null]` so that `[7, 130]` stacks nicely. But if we remove `[7, null]` then the `[7, 130]` bar is stacked floating in the air? That does not seem correct or predictable.

@th1under @Xalio08 I see that the use case you needed is the same use case I have. Did you end up finding solutions?

@100pah 

> Currently, stack is only supported on 'category' axis.

Is there any update on this? Is this still true in 2024? If so, can the documentation be updated? If it is only supported on 'category' axis can you please explain why the specific example of @helgasoft works?

**Comment by helgasoft** [comment]
@kianj2, @gggdomi, I apologize for being too "protective" of ECharts. 
And I admit the data preprocessing workaround in this case is not very practical. Now I agree with you that users should be able to just dump two or more time series and see where bars get stacked. That would be a valuable new feature.
My change of camp was also helped by seeing how [Highcharts supports that already](https://echarts.apache.org/examples/en/editor.html?c=line-simple&code=CQOg7gFgpgdgFAKAARNAcygFwMoGMBOAlgA6aIooDkEmmxAzgFwD0zuAJjCAFb3tQAbQgDd8IGFmYxiAW2YRCaCLggBDfJnoABAIw6QAFhAAOeYuVqN9EDMJdelZEgCUCZyHYB7CXABmAVxhcTEJvJDhnJABvJwAJcxV1TRB6LAB5UlCYejgYiiQ0AU8AI1UBRminfP9UgFUAFQBhCt8y1KqkAF8nTucAbgQnYXUkT0zveiQAXkr8xI0KvPykfFh-fHrPCsp5zABaXG9MVTsofEoAGg6UAC9PTxl6gE9iKG2AD0vrpHpiVVw7GgAErmTAVABMAAYrsskJgXm8kDtPAJ_DIYI58p0YRQQpgBIilvlMFB3mCkQApEpIAAKnnoIRgaEmxSeSAAIoQGURiv4Qt5MRRsR13gBBd5cxbfPEEqWwlAksnbdmqEmCrE44kI7bsVVYQgyKDqrqapBPcWS2bLGWE74K0nkyhU4q0-mM5nGlDdDUdYhFTAZfnZOXLVJEKBMK2whn_ADWgO2ME8-BkZU9XQ6wvyYcIEYqAG1vkTljBVIblYRcEbTfldccC3aKPmdABmcEAdmMAE4uwA2SEDgcXJAAVgAujXlqxm23213jP3B0OkDB_AIBBPGygZ-DjJD2-CDEvoUh25v5U3W7udCPjHul8Pe2PG8_YVnlsX8qXy0jGuooGcSAAEL-IQAjrF8F51qoDYXtuV6dj2i6DsOLbnnBO5zgux4nk-k75NOV57geR4PqO6EXjuxg3nex7DgYr7yoxWIdIxnQDCgwz4EguzTCuUBgEg8RKLs1iNJYZBjEG9D9AgvR9EAA).

![image](https://github.com/user-attachments/assets/ffbd31d6-124c-4261-af4a-7981485c4bac)

**Comment by kianj2** [comment]
@helgasoft No worries 😄 I've seen you comment on a couple issues in this repo and it is always nice to see your input.

Good example. Yes it would be nice if echarts could have the same functionality.

With the preprocessing solution, do you know if this is the intended way to use stacked bars on a time axis? Or is it something that just in your experience has worked ok? I ask because my team needs a solution and I am trying to figure out how reliable the preprocessing one is.

**Comment by helgasoft** [comment]
> this is the intended way to use stacked bars on a time axis? Or is it something that just in your experience has worked ok? 

Would presume there is no "intended way" since the official API does not mention it explicitly and no examples.   
I rarely dare to look at ECharts source code. But as experienced user I trust the workaround because it makes time axis behave like category axis. And category axis is definitely supported.

**Comment by davidchalifoux** [comment]
+1 for this feature.
I'd rather not have to pad my data unnecessarily with empty values.
